pageserver/src/tenant/timeline/compaction.rs

```diff
@@ -668,10 +668,24 @@ impl Timeline {
 
         // Define partitioning schema if needed
 
+        let l0_l1_boundary_lsn = {
+            // We do the repartition on the L0-L1 boundary. All data below the boundary
+            // are compacted by L0 with low read amplification, thus making the `repartition`
+            // function run fast.
+            let guard = self.layers.read().await;
+            let l0_min_lsn = guard
+                .layer_map()?
+                .level0_deltas()
+                .iter()
+                .map(|l| l.get_lsn_range().start)
+                .min()
+                .unwrap_or(self.get_last_record_lsn());
+            l0_min_lsn.max(self.get_ancestor_lsn())
+        };
         // FIXME: the match should only cover repartitioning, not the next steps
         let (partition_count, has_pending_tasks) = match self
             .repartition(
-                self.get_last_record_lsn(),
+                l0_l1_boundary_lsn,
```

Comment:
This LSN is exclusive, right? In other words, we're ignoring the passed LSN, but reading anything below it? Is `disk_consistent_lsn` also exclusive, in case we switch to it?

My understanding is that the Pageserver generally uses exclusive LSNs, just checking.

Reply:
I think I'll need to -1 here, otherwise it will still read one L0 layer.

Reply:
On a second thought I'd like to keep things simple here. Even if we are off by 1 we only need to read one extra L0. But if we -1, we need to think about LSN alignment things which I'm not quite sure about right now.